drivers: dma_mcux_lpc: Fix some DT issues

decsny opened the pull request:
* remove the sdk based TOTAL_DMA_CHANNELS macro and instead just use the zephyr driver's num_of_channels field,
* use dma-channels DT prop correctly (describe hardware not boot configuration), 
* fix the dma-channels property to be the correct number for platforms that currently have an incorrect value in DT
* add the dma-channels property to "enabled" lpc dma nodes that don't currently have it
* set num of otrigs at SOC level

Commit:
drivers: dma_mcux_lpc: remove SDK based macro

remove the sdk based TOTAL_DMA_CHANNELS macro and instead just use the
zephyr driver's num_of_channels field

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -298,14 +297,8 @@ static int dma_mcux_lpc_configure(const struct device *dev, uint32_t channel,
 		return -EINVAL;
 	}
 
-#if defined FSL_FEATURE_DMA_NUMBER_OF_CHANNELS
```
Doesn't total DMA channels here mean something different than `dev_config->num_of_channels`? If I understand correctly:
- `dev_config->num_of_channels`: the number of channel slots we expect the driver to allocate (can be less than number of channels in HW)
- `total_dma_channels`: the number of channels that are actually present on the hardware.

danieldegrasse replied:
`num-of-channels` sets the number of channel configuration structures that the driver allocates. You can see here: https://github.com/nxp-zephyr/zephyr/blob/fix/remove_lpc_dma_macro/drivers/dma/dma_mcux_lpc.c#L349 that we allocate a "virtual channel" value when a new channel is configured. With this PR, you are effectively requiring that the `number-of-channels` property be equivalent to the number of channels the hardware supports.

Based off the documentation of `number-of-channels`, it looks like this driver was previously misusing this DT property. Given that, I think a change here makes sense. I see two ways we could fix this:

- require that the DMA driver statically allocate `num-of-channels` DMA configuration structures, and remove the logic to handle virtual channels from the driver (so the channel number always matches the index of the channel configuration structure)
- retain the ability to configure the number of statically allocated DMA config structures, but make this property configurable via a Kconfig or another nxp specific devicetree property. This way, we do not misuse this `number-of-channels` devicetree property anymore.

Personally, I would prefer the second approach, but I am open to feedback.

decsny replied:
Or, if it is ever actually needed to pretend there are less dma channels in hardware than there actually are, one can just reduce the number of channels DT property, and everything will work the same way. I think you are over-complicating this.

decsny replied:
Yes, there is an argument for a Kconfig that supports a case where a user might want to use slightly less RAM by allocating less channels than needed, but this is a functionality that in my opinion we don't correctly support yet, so it is outside the scope of the PR. 

The devicetree should describe hardware not software configuration, and that's what this PR is fixing, as well as removing the reliance on SDK macros which are notoriously unstable.

teburd replied:
this should be the count of hardware channels imo, so this is the right approach now

dleach02 replied:
@teburd, if you are okay with this PR can you give it an approval?

Commit:
dts: lpc55S6X: Set DMA num otrigs at SOC level

Set the DMA number of otrigs DT property at the SOC level instead of the
board DTS because it is an SOC property and does not change on different
boards.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
dts: lpc dma: Use dma-channels prop correctly

Current erroneous usages of dma-channels prop by lpc-dma nodes:
* dma-channels devicetree property should describe the number of
  channels supported by the dma controller, not the number of channels
  in use.
* LPC55SXX and RTXXX SOCs should be setting dma channels prop at SOC
  level, not board level, since it is an SOC property, not a board
  property.
* lpc55s28 has 23 channels for dma0, not 20.
* lpc55s28 has 10 channels for dma1, not 0.
* lpc55s69 has 23 channels for dma0, not 20.
* rt5xx has 37 channels for dma1, not 0.
* rt6xx has 33 channels for dma0, not 20.
* rt6xx has 33 channels for dma1, not 0.

Fix all of these issues

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts:
```diff
@@ -399,15 +399,10 @@ arduino_serial: &flexcomm12 {
+	status = "okay";
+};
+
+&dma1 {
```
Why are you enabling `dma1` for the RT595 and RT685 here, if it was not previously enabled?

decsny replied:
why not

danieldegrasse replied:
Are we using this DMA node anywhere? If not then we are needlessly increasing the size of the build for the RT595 EVK

danieldegrasse commented on drivers/dma/dma_mcux_lpc.c:
```diff
@@ -676,25 +669,18 @@ static int dma_mcux_lpc_init(const struct device *dev)
 	 * assigned to store dma channel data
 	 */
-	for (int i = 0; i < total_dma_channels; i++) {
+	for (int i = 0; i < config->num_of_channels; i++) {
```
If we are not going to expose a method to the user to allocate configuration data for less than the number of channels present in hardware, this loop (as well as the code here:  https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/dma_mcux_lpc.c#L356) is no longer required, as we don't need a concept of a "virtual channel".

Are you planning to expose a property to set the number of channel configuration structures available in a subsequent PR? Otherwise, we should remove the "virtual channel" management code from the driver.

decsny replied:
yes I think we should keep the virtual channel stuff for now in case we want to add support for limiting the amount of memory the driver uses based on the channels being used, then we don't have to re-add it, I might do a pr if I have time

mmahadevan108 commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69.dtsi:
```diff
@@ -181,20 +181,6 @@ mikrobus_serial: &flexcomm2 {
-	dma-channels = <20>;
-	nxp,dma-num-of-otrigs = <4>;
-	status = "okay";
-};
```
Do we want to delete everything here or just the portion with `dma-channels`

decsny replied:
it is all being moved to other files

mmahadevan108 commented on boards/arm/lpcxpresso55s69/lpcxpresso55s69_cpu0.dts:
```diff
@@ -136,26 +136,6 @@
-	dma-channels = <10>;
-	nxp,dma-num-of-otrigs = <4>;
-	status = "okay";
-};
```
We are removing `nxp,dma-num-of-otrigs` as well. Should we delete only `dma-channels`

decsny replied:
all of these things are just being relocated
